Code/Unsupervised_reconstruction/data/reconstruction_datasets.py

```diff
@@ -48,4 +53,57 @@ def trainingDataLoader(raw_samples:list[str],
 
 
 #-------------Sampling Datasets---------------------
+Raw_minibatch_type = tuple[tuple[
+                        tuple[Tensor], # tuple of c ByteTensors of shape (b, |x|)
+                        list[dict[ModernLanguages, Tensor]] # list of c dicts with the c cognates group
+                        ],
+                        tuple[int, int] # the coords of the minibatch in the global batch
+                        ]
 
+class __SamplingDataset(IterableDataset[Raw_minibatch_type]):
+    """
+    Returns an iterable of tuple([list of c tuple(b samples tensor, associated cognates)], coords in global batch)
+    """
+    def __init__(self, proposals: list[Tensor], cognates: list[dict[ModernLanguages, Tensor]],
+                 samples_number_per_batch: int, cognates_batch_index:int) -> None:
+        assert(len(proposals)==len(cognates)), "the two lists must have the same length to match together"
+        self.proposals_dataset = TensorDataset(*proposals)
+        self.cognates = cognates
+        self.batch_size = samples_number_per_batch
+        self.cognates_batch_index = cognates_batch_index
+    
+    def __iter__(self) -> Generator[Raw_minibatch_type, Any, None]:
+        worker_info = get_worker_info()
+        worker_id, num_workers = 0, 1
+        if worker_info is not None:
+            worker_id = worker_info.id
+            num_workers = worker_info.num_workers
+        samples_batch_index = worker_id
+        for start_index in range(worker_id*self.batch_size, len(self.proposals_dataset), num_workers*self.batch_size):
+            yield((self.proposals_dataset[start_index:start_index+self.batch_size], self.cognates), (self.cognates_batch_index, samples_batch_index))
```

Comment:
Appears to contain an error in the yield.
(Need more time to be certain.)

Reply:
After re-reading, the expression for _yield_ seems OK. However, the type returned by `self.proposals_dataset[start_index:start_index+self.batch_size]` is a `tuple[Tensor, ...]` and not a `list[Tensor]`.

@Convolutio, can you elaborate on your perspective regarding the use of this expression?

Reply:
The raw proposal mini-batch got from the `TensorDataset` instance with this instruction :
> `self.proposals_dataset[start_index:start_index+self.batch_size]`

is in the format of a tuple of $c$ tensors of shape $(b, |x|)$, with $b$=`self.batch_size`. $c$ is the number of cognate pairs given in the `__SamplingDataset` instance. Afterwards, this tuple is being able to be collated in the `__sampling__collate_fn` function.

`start_index` has been set up so that two distinct workers do not load any common mini-batch.